localstack/services/kinesis/kinesis_mock_server.py

```diff
@@ -0,0 +1,143 @@
+import logging
+from typing import Optional
+
+from localstack import config
+from localstack.services import install
+from localstack.utils.common import (
+    TMP_THREADS,
+    ShellCommandThread,
+    chmod_r,
+    get_free_tcp_port,
+    mkdir,
+)
+from localstack.utils.run import FuncThread
+from localstack.utils.serving import Server
+
+LOG = logging.getLogger(__name__)
+
+
+class KinesisMockServer(Server):
+    """
+    Server abstraction for controlling Kinesis Mock in a separate thread
+    """
+
+    def __init__(
+        self,
+        port: int,
+        bin_path: str,
+        latency: str,
+        host: str = "localhost",
+        log_level: str = "INFO",
+        data_dir: Optional[str] = None,
+        initialize_streams: Optional[str] = None,
+    ) -> None:
+        self._latency = latency
+        self._initialize_streams = initialize_streams
+        self._data_dir = data_dir
+        self._bin_path = bin_path
+        self._log_level = log_level
+        super().__init__(port, host)
+
+    def do_start_thread(self) -> FuncThread:
+        cmd = self._create_shell_command()
+        LOG.debug("starting kinesis process %s", cmd)
+        t = ShellCommandThread(
+            cmd,
+            strip_color=True,
+            log_listener=self._log_listener,
+            auto_restart=True,
+        )
+        TMP_THREADS.append(t)
+        t.start()
+        return t
+
+    def _create_shell_command(self) -> str:
+        """
+        helper method for creating kinesis mock invocation command
+        """
+        if self._data_dir:
+            kinesis_data_dir_param = "SHOULD_PERSIST_DATA=true PERSIST_PATH=%s" % self._data_dir
+        else:
+            kinesis_data_dir_param = ""
+        log_level_param = "LOG_LEVEL=%s" % self._log_level
+        latency_param = (
+            "CREATE_STREAM_DURATION={l} DELETE_STREAM_DURATION={l} REGISTER_STREAM_CONSUMER_DURATION={l} "
+            "START_STREAM_ENCRYPTION_DURATION={l} STOP_STREAM_ENCRYPTION_DURATION={l} "
+            "DEREGISTER_STREAM_CONSUMER_DURATION={l} MERGE_SHARDS_DURATION={l} SPLIT_SHARD_DURATION={l} "
+            "UPDATE_SHARD_COUNT_DURATION={l}"
+        ).format(l=self._latency)
+        init_streams_param = (
+            "INITIALIZE_STREAMS=%s" % self._initialize_streams if self._initialize_streams else ""
+        )
+
+        if self._bin_path.endswith(".jar"):
+            cmd = (
+                "KINESIS_MOCK_PLAIN_PORT=%s SHARD_LIMIT=%s %s %s %s %s java -XX:+UseG1GC -jar %s"
+                % (
+                    self.port,
+                    config.KINESIS_SHARD_LIMIT,
+                    latency_param,
+                    kinesis_data_dir_param,
+                    log_level_param,
+                    init_streams_param,
+                    self._bin_path,
+                )
+            )
+        else:
+            chmod_r(self._bin_path, 0o777)
+            cmd = "KINESIS_MOCK_PLAIN_PORT=%s SHARD_LIMIT=%s %s %s %s %s %s --gc=G1" % (
```

Comment:
There's lots of string interpolation going on here and in the `kinesalite` server as well. I personally find the `%s` syntax hard to follow and would like to create a non-urgent issue for refactoring these instances to use `.format()` instead. Thoughts?

Reply:
Great point. The `%s` interpolation syntax is mostly for historical reasons (in the past we used to be compatible with python 2 & 3) - these days we tend to use f-strings (or `.format()`) instead. 👍